clang/lib/Sema/SemaAttr.cpp

```diff
@@ -216,6 +216,59 @@ void Sema::inferGslOwnerPointerAttribute(CXXRecordDecl *Record) {
   inferGslPointerAttribute(Record, Record);
 }
 
+void Sema::inferLifetimeBoundAttribute(FunctionDecl *FD) {
+  if (FD->getNumParams() == 0)
+    return;
+
+  if (unsigned BuiltinID = FD->getBuiltinID()) {
+    // Add lifetime attribute to std::move, std::fowrard et al.
+    switch (BuiltinID) {
+    case Builtin::BIaddressof:
+    case Builtin::BI__addressof:
+    case Builtin::BI__builtin_addressof:
+    case Builtin::BIas_const:
```

Comment:
I start to have the feeling we are writing code like this over and over again.
Some examples:
https://github.com/llvm/llvm-project/blob/df57833ea8a3f527b7941576b4a130ddd4361e61/clang/lib/Analysis/BodyFarm.cpp#L718
https://github.com/llvm/llvm-project/blob/df57833ea8a3f527b7941576b4a130ddd4361e61/clang/lib/AST/ExprConstant.cpp#L8763

But there are more, and I always wonder if they can get out of sync. I wonder if this is time to introduce some helper functions that can be reused.

Reply:
(The latter two cases are identical, we could potentially abstract a common function.)

I think this should be handled on a case-by-case basis, depending on the specific logic involved.  It seems fine to me have this small amount of duplication.